feat: consensus key rotation of validator (#15165)
cosmos/cosmos-sdk: x/staking/README.md

https://github.com/cosmos/cosmos-sdk/blob/v0.47.0-rc1/proto/cosmos/staking/v1beta1/staking.proto#L263-L308
```

## ConsPubkeyRotation

> The `ConsPubkey` of a validator will be instantly rotated to the new `ConsPubkey` there will be a tracking for the rotations which is only allows limited no.of rotations should be made in the unbonding period of time

Reviewer: ```suggestion
The `ConsPubkey` of a validator will be instantly rotated to the new `ConsPubkey`. The rotation will be tracked to only allow a limited number of rotations within an unbonding period of time.
```

`ConsPubkeyRotation` are indexed in the store as:

ConsPubKeyRotationHistory: `0x70 | validatorAddrLen(1 byte) | validatorAddr | rotated_height -> ProtocalBuffer(ConsPubKeyRotationHistory)`
ValidatorConsensusKeyRotationRecordQueue: `0x72 | format(time) -> ProtocalBuffer(RotatedValAddrs)`

ValidatorConsensusKeyRotationRecordIndex:`0x73 | valAddressLen(1 byte) | valAddr | format(time) -> ProtocalBuffer([]Byte{})`

`ConsPubKeyRotationHistory` is used for querying the rotations of a validator

`ValidatorConsensusKeyRotationRecordQueue` is to keep track of the rotation across the unbonding period (waiting period in the queue), this be pruned after the unbonding period of waiting time.

`ValidatorConsensusKeyRotationRecordIndex` is to keep track of a validator that how many rotations were made inside unbonding period. This be pruned after the unbonding period of waiting time.

A `ConsPubKeyRotationHistory` object is created every time a consensus pubkey rotation occurs.

To prevent the spam: 

* There will only limited number of rotations can be done within unbonding period of time. 

> * And significant amount of fee will be deducted for rotating the keys.

Reviewer: ```suggestion
* A non-negligible fee will be deducted for rotating a consensus key.
```

Formula to calculate the fees to rotate the key is: 

```go

> (max(VotingPowerPercentage 100, 1) * InitialKeyRotationFee) * 2^(number of rotations in ConsPubKeyRotationHistory in recent unbonding period)

Reviewer: What is `VotingPowerPercentage 100`?

Author: my bad, I think this must be VotingPowerRatio.

Author: I think it should just be `VotingPowerPercentage`

* remove the entry from the `Redelegation` object

#### Consensus pubkey rotation

> When a `ConsPubkeyRotation` occurs the validator and the `ValidatorConsensusKeyRotationRecordQueue` updated

Reviewer 2: ```suggestion
When a `ConsPubkeyRotation` occurs the validator and the `ValidatorConsensusKeyRotationRecordQueue` are updated
```

> * the old consensus pubkey address will be removed from state and new consensus pubkey address will be added in place.

Author: No

> * the old consensus pubkey address will be removed from state and new consensus pubkey address will be added in place.

Reviewer 2: if a validator misbehaves and swaps their consensus key we should still know how to slash the validator if evidence comes in after the rotation

> * the old consensus pubkey address will be removed from state and new consensus pubkey address will be added in place.

Reviewer 2: seems like this is already done. we just need to modify how evidence is tracked a bit i belive

* remove the `Redelegation` object from the store if there are no
  remaining entries.

#### ConsPubKeyRotations

> After waiting for unbonding period of time the matured rotations will be removed from queues and indexes to unblock the validator for the next iterations

Author: Yes
